shared/src/Unison/TermParser.hs

```diff
@@ -0,0 +1,239 @@
+module Unison.TermParser where
+
+import Prelude hiding (takeWhile)
+
+import Control.Applicative
+import Data.Char (isDigit, isAlphaNum, isSpace, isSymbol, isPunctuation)
+import Data.Foldable (asum)
+import Data.Functor (($>), void)
+import Data.List (foldl')
+import Data.Set (Set)
+import Unison.Parser
+import Unison.Symbol (Symbol, Symbol(..))
+import Unison.Term (Term, Literal)
+import Unison.Type (Type)
+import Unison.View (DFO)
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Unison.Term as Term
+import qualified Unison.TypeParser as TypeParser
+import qualified Unison.Var as Var
+
+type V = Symbol DFO
+
+term :: Parser (Term V)
+term = possiblyAnnotated term2
+
+term2 :: Parser (Term V)
+term2 = let_ term3 <|> term3
+
+term3 :: Parser (Term V)
+term3 = infixApp term4 <|> term4
+
+infixApp :: Parser (Term V) -> Parser (Term V)
+infixApp p = f <$> arg <*> some ((,) <$> infixVar <*> arg)
+  where
+    arg = p
+    f :: Term V -> [(V, Term V)] -> Term V
+    f = foldl' g
+    g :: Term V -> (V, Term V) -> Term V
+    g lhs (op, rhs) = Term.apps (Term.var op) [lhs,rhs]
+
+term4 :: Parser (Term V)
+term4 = prefixApp term5
+
+term5 :: Parser (Term V)
+term5 = lam term <|> termLeaf
+
+termLeaf :: Parser (Term V)
+termLeaf = asum [prefixTerm, lit, parenthesized term, blank, vector term]
+
```

Comment:
These are really more notes to self that were useful when you were writing. I'd probably delete and just have as a note at the top:

> Precedence of language constructs is identical to Haskell, except that all operators (like `+`, `<*>`, or any sequence of non-alphanumeric characters) are left-associative and equal precedence, and operators must have surrounding whitespace (`a + b`, not `a+b`) to distinguish from identifiers that may contain operator characters (like `empty?` or `fold-left`).

This is economical and unambiguous.